### PR TITLE
feat: strict basedpyright, finish EqualityConstraint rename, agentic docs

### PR DESCRIPTION
**Pull request by antonioterpin**
## Summary

Port flowbench-sw's stricter typing conventions to pinet and finish a
half-applied API rename that was breaking 217/572 tests at runtime.

- **Strict basedpyright.** Drop the permissive `[tool.pyright]` block and
  run the project-wide type check (`pass_filenames: false`) with
  `--level error` — matches the flowbench-sw invocation. `src/` *and*
  `src/test/` both type-check now. 0 errors across the whole tree.
- **Complete the `A → a_dyn` / `C → constr_matrix` rename.** Library and
  tests were calling `EqualityConstraint(A=..., var_A=...)` and
  `AffineInequalityConstraint(C=...)`, which fail at runtime against the
  post-rename API. 217/572 tests were red on this branch pre-PR; all 572
  pass now.
- **Fix `ConstraintParser` dispatch.** `__init__` was assigning
  `self.parse = self.parse_polytope`, but `parse_polytope` never
  existed (the body lived on a method named `parse`). Any non-identity
  construction raised `AttributeError`. Introduced an explicit
  `parse()` dispatcher that routes to `parse_polytope` /
  `parse_non_linear` / an identity closure.
- **Dataclass bugs caught by strict mode.** `Optional` was used without
  being imported (latent `NameError` in validators); `EXPECTED_BOUND_NDIM`
  / `EXPECTED_PROJECTION_NDIM` were referenced but never defined.
- **Ruff + pydoclint config** updated to recognise the math-heavy
  identifier convention (matrix names `A`, `A_soc_1`, `kA`, `Cx`, …) and
  to scope pydoclint to library code only.
- **Agentic docs.** New `.agent/`, `.claude/`, and `docs/` mirroring the
  flowbench-sw layout (adapted — no hardware/goggles content) so
  Claude Code, Antigravity, and custom agents share the same workflows.

### Before / after

| Gate                               | Before          | After       |
|------------------------------------|-----------------|-------------|
| `uv run pytest`                    | 355 / 572 pass  | **572 / 572** |
| `uv run basedpyright`              | 287 errors      | **0 errors** |
| `uv run pre-commit run --all-files` | ruff + pydoclint + basedpyright fail | **all pass** |

## Commits

1. `docs: import agent-driven development documentation`
2. `chore(tooling): tighten typing and linting to match flowbench-sw conventions`
3. `fix(pinet): complete EqualityConstraint API rename and ConstraintParser dispatch`
4. `test: update suite to the new API and add type narrowing`
5. `chore(benchmarks): add generic type args and repair loop-bound variables`
6. `chore(tools): type the autotuning DataLoader cast and load_configuration return`

## Test plan

- [x] `uv run pytest` — 572 / 572 pass
- [x] `uv run pre-commit run --all-files` — ruff, ruff-format, pydoclint, basedpyright, yaml/eof/whitespace all pass
- [x] `uv run pre-commit run --hook-stage push --all-files` — same gates green at push stage
- [ ] Spot-check one benchmark run (`uv run python -m benchmarks.QP.run_qp ...`) to confirm no regression in the dispatch refactor on the scripts path
- [ ] CI: confirm the `Coding style validation` and `Python CI` workflows go green once `uv` is available in the runner (separate issue — CI currently installs `pre-commit` via `pip` only, so `uv run basedpyright` short-circuits)

## Notes for the reviewer

- The ConstraintParser refactor is the largest diff in commit 3 — the
  polytope/non_linear paths are now typed but the math is unchanged.
  Worth eyeballing the block that builds `a_dyn_lifted` / `b_lifted` in
  `parse_polytope` and the auxiliary-variable construction in
  `parse_non_linear`.
- `CartesianConstraint.__init__` now takes `list[SocConstraint] | None`
  instead of `list[NonLinearConstraint] | None` because the validator
  only accepts SocConstraint. If other non-linear constraint types land
  later, widen the signature back out.
- Two dead gradient-check blocks referencing `loss` / `lossvmapped` from
  a different scope were removed from `test_vjp.py`. They were
  unreachable before (the parser raised earlier) and exposed a
  `NameError` once the parser started working.
- CI note: the `pre-commit` hook for `basedpyright` now uses
  `pass_filenames: false` and `--level error`. If the tolerance on
  warnings later matters (e.g. `reportAny` reducing), flip `--level` to
  `warning` to re-introduce the advisory gate.

### CODE REVIEW

File: src/benchmarks/toy_MPC/load_toy_mpc.py [hunk 1]
```diff
@@ -130,9 +135,34 @@ def __init__(
...
+        """
+        for start in range(0, len(self.dataset), self.batch_size):
+            batch_idx = self._perm[start : start + self.batch_size]
+            yield self.dataset[batch_idx]
+        if self.shuffle:
+            self._perm = self._get_perm()
+
+    def _get_perm(self) -> jax.Array:
+        self.rng_key, last_key = jax.random.split(self._rng_key)
+        return jax.random.permutation(last_key, len(self.dataset))
```
[the remark this answers is not in the document]
Comment (antonioterpin) [reply src/benchmarks/toy_MPC/load_toy_mpc.py:180]: Fixed in 5f5f15f. Changed `_get_perm` to update `self._rng_key` in place (and dropped the now-unused public `self.rng_key` alias). Applied the same fix to the QP loader, which had the identical bug.

File: src/pinet/dataclasses.py [hunk 2]
```diff
@@ -262,6 +290,8 @@ def to_nl_spec(self) -> "NonLinearSpecification":
             NonLinearSpecification: A NonLinearSpecification instance with SOCType
                 and the constraints from this specification.
         """
+        if self.mask_t is None:
+            raise ValueError("mask_t must be set to convert to NonLinearSpecification.")
         return NonLinearSpecification(
```
[the remark this answers is not in the document]
Comment (antonioterpin) [reply src/pinet/dataclasses.py]: Fixed in 5f5f15f. Replaced the ad-hoc `mask_t is None` check with `self.validate()`, which enforces both masks (and the shape/dtype invariants) before constructing the NL spec.

File: src/tools/utils.py [hunk 3]
```diff
@@ -4,14 +4,13 @@
...
-import yaml
-
 import wandb
+import yaml
 
 logger = logging.getLogger(__name__)
 
 
-def load_configuration(file_path: str) -> dict:
+def load_configuration(file_path: str) -> dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (antonioterpin) [reply src/tools/utils.py:13]: Fixed in 5f5f15f. Added an `isinstance(..., dict)` check that raises `TypeError` when `yaml.safe_load` returns `None` or a scalar, so the `dict[str, Any]` contract is enforced at the boundary.

File: docs/workflows/refactor.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+description: Code refactoring without behavior change
+---
+
+Use when improving structure, naming, or internal APIs without changing outputs.
+
+1. Identify the refactor scope and keep it small and coherent.
+2. Ensure existing tests cover the behavior.
+   - If coverage is weak, add characterization tests first.
+3. Apply the refactor in small steps (prefer multiple local commits).
+// turbo
```
[the remark this answers is not in the document]
Comment (antonioterpin) [reply docs/workflows/refactor.md:11]: Dropping. `// turbo` is an agent-directive marker (auto-approve step) inherited verbatim from the flowbench-sw `docs/workflows/` files — I want the two projects to stay byte-identical so the same agents can consume either. It's not meant as markdown prose; it's a literal token the agent harness looks for. If we ever render these workflows as user-facing docs, I'll switch both projects to `<!-- turbo -->` together.

File: docs/workflows/feature.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+4. Run formatting/lint gate:
+   ```bash
+   uv run pre-commit run --all-files
+   ```
+// turbo
+5. Run tests to confirm they pass:
+   ```bash
+   uv run pytest
+   ```
+// turbo
```
[the remark this answers is not in the document]
Comment (antonioterpin) [reply docs/workflows/feature.md:24]: Dropping. `// turbo` is an agent-directive marker (auto-approve step) inherited verbatim from the flowbench-sw `docs/workflows/` files — I want the two projects to stay byte-identical so the same agents can consume either. It's not meant as markdown prose; it's a literal token the agent harness looks for. If we ever render these workflows as user-facing docs, I'll switch both projects to `<!-- turbo -->` together.

File: docs/workflows/api-validation.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+6. Run formatting/lint gate:
+   ```bash
+   uv run pre-commit run --all-files
+   ```
+// turbo
+7. Run all tests (including the new ones):
+   ```bash
+   uv run pytest
+   ```
+// turbo
```
[the remark this answers is not in the document]
Comment (antonioterpin) [reply docs/workflows/api-validation.md:29]: Dropping. `// turbo` is an agent-directive marker (auto-approve step) inherited verbatim from the flowbench-sw `docs/workflows/` files — I want the two projects to stay byte-identical so the same agents can consume either. It's not meant as markdown prose; it's a literal token the agent harness looks for. If we ever render these workflows as user-facing docs, I'll switch both projects to `<!-- turbo -->` together.

File: docs/workflows/bugfix.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+---
+description: Bugfix or regression workflow
+---
+
+Use when fixing a bug or preventing a known regression.
+
+1. Write a failing test that reproduces the issue (minimal reproduction).
+2. Implement the fix in the correct module under `src/pinet/`.
+// turbo
```
[the remark this answers is not in the document]
Comment (antonioterpin) [reply docs/workflows/bugfix.md:9]: Dropping. `// turbo` is an agent-directive marker (auto-approve step) inherited verbatim from the flowbench-sw `docs/workflows/` files — I want the two projects to stay byte-identical so the same agents can consume either. It's not meant as markdown prose; it's a literal token the agent harness looks for. If we ever render these workflows as user-facing docs, I'll switch both projects to `<!-- turbo -->` together.

File: docs/workflows/docs.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+Use when updating docstrings, README/docs, or comments.
+
+### Standards
+- Adhere to **Docstring & Typing** rules in [`typing-docstrings.md`](../standards/typing-docstrings.md).
+- Follow **Markdown heading capitalization** in [`code-clarity.md`](../standards/code-clarity.md#header-and-title-capitalization): use sentence case (capitalize only the first word, except proper nouns/acronyms).
+- Follow **Markdown link path rules** in [`code-clarity.md`](../standards/code-clarity.md#markdown-link-paths): use relative links for repository-internal targets.
+- Focus on "why" (rationale) rather than just "what" (implementation).
+
+1. Make doc changes (focus on "why" and "how to use").
+// turbo
```
[the remark this answers is not in the document]
Comment (antonioterpin) [reply docs/workflows/docs.md:14]: Dropping. `// turbo` is an agent-directive marker (auto-approve step) inherited verbatim from the flowbench-sw `docs/workflows/` files — I want the two projects to stay byte-identical so the same agents can consume either. It's not meant as markdown prose; it's a literal token the agent harness looks for. If we ever render these workflows as user-facing docs, I'll switch both projects to `<!-- turbo -->` together.

**Comment by mattiademartino** [comment]
I ran `basedpyright --level error` and `pre-commit run --all-files`.

Most checks are passing (`ruff`, `ruff format`, `pydoclint`, `check yaml`, `fix end of files`, and `trim trailing whitespace`), but the `basedpyright` hook is currently failing with two errors:

1. In `src/benchmarks/other_projections.py:69`, `basedpyright` seems to expect a class, but it receives a callable/function instead.

2. In `src/test/test_equilibration.py:124`, the inferred type is `list[Array | Inequality | Constraint]`, which is not assignable to `list[Constraint]` because Python lists are invariant. The error message suggests using `Sequence[Constraint]` instead.

**Comment by antonioterpin** [comment]
@mattiademartino thanks for the repro! Those two errors are Python-3.10-specific — `uv.lock` resolves cvxpy to **1.7.5** on 3.10 (marker `python_full_version < '3.11'`) and to **1.8.2** on 3.11+. The 1.7.5 stubs are what produce both diagnostics: `CvxpyLayer` resolves to a function in cvxpylayers' 1.7.5-compatible typings, and `cp.problem.constraints` is typed `list[Array | Inequality | Constraint]` rather than the narrower `list[Constraint]` in 1.8.2.

The `Coding style validation` workflow on this branch now runs on Python **3.13** (commit c0ded37 / b798dc5 bumped it) so basedpyright sees the same 1.8.2 stubs contributors run locally. The `Python CI` (test) workflow keeps 3.10 to verify the library's declared minimum.

tl;dr: basedpyright should be clean if you run it with Python 3.11+ in the venv. If you want to keep 3.10 locally, you can either (a) pin cvxpy to 1.8.2 in your venv via `uv sync --python 3.11+`, or (b) tolerate those two diagnostics until we decide whether 3.10 is still worth supporting as a typed target.

**Comment by antonioterpin** [comment]
Independent pass on the diff after Copilot's review. Most of what I'd have flagged is already covered by the existing inline thread or by your basedpyright follow-up — one stale-naming straggler is the only thing I caught:

- **`src/pinet/constraints/constraint_parser.py`** — search for "lifted A" in the comments inside `parse_polytope` / `parse_non_linear`. At least one block-construction comment still reads "Build first block column of lifted A" but should be "lifted a_dyn" to match the new naming convention. Functionally harmless; just inconsistent with the rename this PR finishes.

The rest checks out:

- Constraint-parser dispatch: `_identity_mode` is locked at construction, the three branches in `parse()` are mutually exclusive, no shared state mutated between methods. Safe.
- `affine_equality.py`: I grepped for stray `\.C` / `var_A` / `\.A\b` references in `src/pinet/`. None outside `_A`/`a_dyn` machinery and the obvious matrix-multiplication math identifiers (which the docs/standards/code-clarity.md exemption explicitly allows). Rename is complete.
- `pyproject.toml` `[tool.basedpyright]` block: all three disables (`reportMissingSuperCall`, `reportImplicitAbstractClass`, `reportConstantRedefinition`) check out against the cited reasons — eqx.Module's `__init__` doesn't super(), eqx enforces abstract methods natively, and the math-identifier `_A`/`_B`/`Cx` convention is documented in `docs/standards/code-clarity.md`.
- The new `docs/` standards/workflows: sampled `code-quality.md`, `feature.md`, `api-validation.md`, `refactor.md`. All commands (`uv run pytest`, `uv run pre-commit run --hook-stage push --all-files`) are real, paths (`src/pinet/`, `src/test/`, `docs/standards/`) all exist.

LGTM modulo the comment-text touch-up.
